ctypesgen/printer_python/printer.py

```diff
@@ -229,7 +229,35 @@ def print_typedef(self, typedef):
     def print_struct(self, struct):
         self.srcinfo(struct.src)
         base = {"union": "Union", "struct": "Structure"}[struct.variety]
-        self.file.write("class %s_%s(%s):\n" "    pass\n" % (struct.variety, struct.tag, base))
+        if not struct.opaque:
+            self.file.write("@struct_decorator\n")
+        self.file.write("class %s_%s(%s):\n" % (struct.variety, struct.tag, base))
+        if struct.opaque:
+            self.file.write("    pass\n")
+        else:
+            # handle unnamed fields.
+            unnamed_fields = []
+            names = set([x[0] for x in struct.members])
+            anon_prefix = "unnamed_"
+            n = 1
+            for mi in range(len(struct.members)):
+                mem = list(struct.members[mi])
+                if mem[0] is None:
+                    while True:
+                        name = "%s%i" % (anon_prefix, n)
+                        n += 1
+                        if name not in names:
+                            break
+                    mem[0] = name
+                    names.add(name)
+                    if type(mem[1]) is CtypesStruct:
+                        unnamed_fields.append(name)
+                    struct.members[mi] = mem
+            struct.unnamed_fields = unnamed_fields
+
+            # print annotations
+            for name, ctype in struct.members:
+                self.file.write(f"    {name} : {ctype.py_string()}\n")
```

Comment:
I think this will break for struct fields whose type references the struct type itself. (There are examples for this in pdfium, where a struct has a field for a callback function which receives a pointer to the struct itself as first argument.) Fields must be defined below the class body for this reason, and so would annotations.

Reply:
This should be handled by the `__future__.annotations` import. If depending on this has drawbacks, we could also explicitly quote types.
I think `Self` is overkill here. If I'm not mistaken, it is intended for inheritance, where a method should return an instance of the derived class.

Reply:
Ah, indeed. Python handles this on its own with the future import.
```pycon
>>> from __future__ import annotations
>>> class Test:
...     selfref: Test
... 
>>> Test.__annotations__
{'selfref': 'Test'}
```

Note that this automatic transform to string impacts how we can reuse annotations, though - e.g. I don't see how we could build fields from annotations like this.

Reply:
And you're right of course, Self is meant for an instance, not the class type itself. Sorry for the blunder.

Reply:
> Note that this automatic transform to string impacts how we can reuse annotations, though - e.g. I don't see how we could build fields from annotations like this.

For a potential solution see https://github.com/ctypesgen/ctypesgen/pull/185#discussion_r1390466448

> But maybe it is doable the other way round, building annotations from fields below the class body?

I doubt it, since (afaik) for annotations to work for IDEs you have to spell them out.

Reply:
> > But maybe it is doable the other way round, building annotations from fields below the class body?

> I doubt it, since (afaik) for annotations to work for IDEs you have to spell them out.

In that case, I would prefer if this were made optional.

Reply:
I trade slots for annotation :sweat_smile: On a more constructive note, using `get_type_hints`, we could probably directly merge the [extended version](https://github.com/zwergziege/ctypesgen/tree/reduce_struct_field_duplication) of this branch. Sounds good?